### PR TITLE
auth-backend: migrate onelogin and auth0 to sign-in resolver

### LINKED ISSUE #6245: auth-backend: implement auth handler and sign-in resolvers for more auth providers
## Background

The Google auth provider has adopted a new pattern for making the authentication more configurable in #4532. The PR added a way to configure both an auth handler and a sign-in resolver, with both of them being callbacks that are called at different points of authentication.

The auth handler is responsible for validating each login and generating a lightweight profile for the user which is used in the frontend only to be showed to the user themselves so that they know what account they are logged in with. The auth handler can also throw an error in order to completely reject the authentication.

The sign-in resolver is responsible for assigning a Backstage Identity to the user. Long-term this will be an optional step that is only implemented for the auth providers that integrators want to use to identify users in Backstage, as opposed to other providers that may only be used for access to third party resources.

You can read more about the auth handlers and sign-in resolvers in the [documentation](https://backstage.io/docs/auth/identity-resolver).

## Future Work

The pattern established in PR #4532 is something that we want to apply to all our current authentication providers, in order to provide more power to the configuration of all different auth providers.

The following is a list of all auth providers and whether they have been refactored or not:

- [x] Auth0 #8658
- [x] AWS-ALB #7380
- [x] GitHub #6912
- [x] GitLab #6535
- [x] Google #4532
- [x] Microsoft #6430
- [x] OAuth2 #6512
- [x] OIDC #8318
- [x] Okta #6713
- [x] OneLogin #8658
- [x] SAML2 #8331

The maintainers will work to migrate these providers over time, but we are thankful for any contributions that help speed up the process 🙏 

## Refactoring

In order to make contributing easier these are some pointers to guide the process. While PR #4532 serves as a template, it may not be easy to see what work was done and why. When migrating a provider, the main chunk if not all of the changes happen in the provider file itself, meaning that if you're looking at the PR it's the changes inside `packages/auth-backend/src/providers/google/` that are the most interesting.

The refactoring can be broken down into these steps:

- The `create*Provider` method should already have an `options` parameter that is probably unused, add the `authHandler` and `signIn.resolver` options to it in the same way as is done in the Google provider. The type parameters of the two options should match for example the first type parameter of `executeFrameHandlerStrategy`, for OAuth-based providers it's `OAuthResult`.
- Refactor the provider class so that the auth result is handled in a single place, if this is not already the case, this is the `handleResult` method in #4532. If the callback passed to the passport strategy is doing a lot of work you will want to refactor it to do as little as possible and move that work out.
- Add `authHandler` and `signInResolver` options to the provider class and use these to transform the auth result, typically replacing things like `makeProfileInto`, `this.tokenIssuer.issueToken` calls. You will likely also need to add new members to the class to be passed on in the context of the `signInResolver` call. Note that the `signInResolver` should be optional and only be used to populate the `backstageIdentity` in the result if it is defined.
- Move the old auth handling and sign-in resolution logic into new functions that we'll use as the default implementations, these are the `googleDefaultSignInResolver` and inline `makeProfileInfo` call in the Google auth provider. Note that any existing sign-in logic is likely sub-optimal, but we keep it in place as is and make it the default in order to avoid a breaking change.
- Refactor the `create*Provider` implementation to pass on the new options and fall back to the defaults, similar to how `createGoogleProvider` was changed in #4532.
- Update or add tests! See #4532 for how to add some basic tests, and don't hesitate to move this step earlier if you want to.
- If there are any obvious additional sign-in resolvers that could be exported along with provider, implement and export these as part of the provider. An example of this in #4532 is the `googleEmailSignInResolver`.
- Chime in in the comments if I missed anything 😉

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Closes #6245

These are the final two providers to be migrated to the sign-in resolver pattern. Quick and dirty without tests because I wanna get them migrated for further refactoring x)

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/auth-backend/src/providers/onelogin/provider.ts [hunk 1]
```diff
@@ -124,51 +137,124 @@ export class OneLoginProvider implements OAuthHandlers {
...
+  info,
+  ctx,
+) => {
+  const { profile } = info;
 
+  if (!profile.email) {
+    throw new Error('OIDC profile contained no email');
+  }
+
+  const id = profile.email.split('@')[0];
```
Comment (freben) [inline plugins/auth-backend/src/providers/onelogin/provider.ts:189]: didn't we say that sub is a full entity ref? (same in the other one of course)

File: plugins/auth-backend/api-report.md [hunk 2]
```diff
@@ -47,6 +47,14 @@ export type AtlassianProviderOptions = {
   };
 };
 
+// @public (undocumented)
```
Comment (freben) [inline plugins/auth-backend/api-report.md:50]: might as well slap the docs on these while we're at it? :)

**Review by freben — APPROVED** [review]
Nice! 🥳  Two comments but happily approved to unblock.

File: plugins/auth-backend/api-report.md (hunk 2 above)
Comment (Rugvip) [reply plugins/auth-backend/api-report.md:50]: Eh, none of them have it so would rather deal with all of them at once at some other point x)

File: plugins/auth-backend/src/providers/onelogin/provider.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/auth-backend/src/providers/onelogin/provider.ts:186]: Oh yeah totally, switched around to let the adapter handle the token creation

**Pull request merged**